File: CS/ControlTemplateGallerySample/ControlTemplateGallery/ControlTemplateStorage.cs
using DevExpress.XtraReports.UI;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace ControlTemplateGallerySample
{
    public interface IControlTemplateStorage
    {
        string[] GetCategoryNames();
        string[] GetTemplateNamesForCategory(string categoryName);
        byte[] GetData(string categoryName, string templateName);
        void SetData(string categoryName, string templateName, byte[] templateLayout);
        void DeleteTemplate(string categoryName, string templateName);
        void DeleteCategory(string categoryName);
    }

    //Default in-memory storage
    public class ControlTemplateStorage : IControlTemplateStorage
    {
        List<TemplateLayoutItem> templates;

        public ControlTemplateStorage()
        {
            templates = new List<TemplateLayoutItem>();            
        }

        internal static XRControl[] GetTemplateControls(IControlTemplateStorage storage, string categoryName, string templateName)
        {
            byte[] layoutBytes = storage.GetData(categoryName, templateName);

            if (layoutBytes != null && layoutBytes.Length > 0)
            {
                XtraReport tempReport = new XtraReport();
                using (MemoryStream stream = new MemoryStream(layoutBytes))
                {
                    tempReport.LoadLayoutFromXml(stream);
                }

                return tempReport.Bands[BandKind.Detail].Controls.Cast<XRControl>().ToArray();
            }

            return null;
        }

        public void DeleteTemplate(string categoryName, string templateName)
        {
            TemplateLayoutItem item = templates.Where(x => x.Category == categoryName && x.Name == templateName).FirstOrDefault();
            templates.Remove(item);
        }

        public void DeleteCategory(string categoryName)
        {
            IEnumerable<TemplateLayoutItem> items = templates.Where(x => x.Category == categoryName);
            
            foreach (TemplateLayoutItem item in items)
                templates.Remove(item);
        }

        public string[] GetCategoryNames()
        {
            return templates.Select(x => x.Category).Distinct().ToArray();
        }

        public string[] GetTemplateNamesForCategory(string categoryName)
        {
            return templates.Where(x => x.Category == categoryName).Select(y => y.Name).ToArray();
        }

        public byte[] GetData(string categoryName, string templateName)
        {
            TemplateLayoutItem item = templates.Where(x => x.Category == categoryName && x.Name == templateName).FirstOrDefault();
            return item == null ? null : item.LayoutBytes;
        }

        public void SetData(string categoryName, string templateName, byte[] templateLayout)
        {
            TemplateLayoutItem item = templates.Where(x => x.Category == categoryName && x.Name == templateName).FirstOrDefault();
            if (item == null)
                templates.Add(new TemplateLayoutItem() { Name = templateName, Category = categoryName, LayoutBytes = templateLayout });
        }        

        class TemplateLayoutItem 
        {
            public string Category { get; set; }
            public string Name { get; set; }
            public byte[] LayoutBytes { get; set; }
        }        
    }
}

File: VB/ControlTemplateGallerySample/ControlTemplateGallery/ControlTemplateStorage.vb
Imports DevExpress.XtraReports.UI
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text

Namespace ControlTemplateGallerySample
    Public Interface IControlTemplateStorage
        Function GetCategoryNames() As String()
        Function GetTemplateNamesForCategory(ByVal categoryName As String) As String()
        Function GetData(ByVal categoryName As String, ByVal templateName As String) As Byte()
        Sub SetData(ByVal categoryName As String, ByVal templateName As String, ByVal templateLayout() As Byte)
        Sub DeleteTemplate(ByVal categoryName As String, ByVal templateName As String)
        Sub DeleteCategory(ByVal categoryName As String)
    End Interface

    'Default in-memory storage
    Public Class ControlTemplateStorage
        Implements IControlTemplateStorage

        Private templates As List(Of TemplateLayoutItem)

        Public Sub New()
            templates = New List(Of TemplateLayoutItem)()
        End Sub

        Friend Shared Function GetTemplateControls(ByVal storage As IControlTemplateStorage, ByVal categoryName As String, ByVal templateName As String) As XRControl()
            Dim layoutBytes() As Byte = storage.GetData(categoryName, templateName)

            If layoutBytes IsNot Nothing AndAlso layoutBytes.Length > 0 Then
                Dim tempReport As New XtraReport()
                Using stream As New MemoryStream(layoutBytes)
                    tempReport.LoadLayoutFromXml(stream)
                End Using

                Return tempReport.Bands(BandKind.Detail).Controls.Cast(Of XRControl)().ToArray()
            End If

            Return Nothing
        End Function

        Public Sub DeleteTemplate(ByVal categoryName As String, ByVal templateName As String) Implements IControlTemplateStorage.DeleteTemplate
            Dim item As TemplateLayoutItem = templates.Where(Function(x) x.Category = categoryName AndAlso x.Name = templateName).FirstOrDefault()
            templates.Remove(item)
        End Sub

        Public Sub DeleteCategory(ByVal categoryName As String) Implements IControlTemplateStorage.DeleteCategory
            Dim items As IEnumerable(Of TemplateLayoutItem) = templates.Where(Function(x) x.Category = categoryName)

            For Each item As TemplateLayoutItem In items
                templates.Remove(item)
            Next item
        End Sub

        Public Function GetCategoryNames() As String() Implements IControlTemplateStorage.GetCategoryNames
            Return templates.Select(Function(x) x.Category).Distinct().ToArray()
        End Function

        Public Function GetTemplateNamesForCategory(ByVal categoryName As String) As String() Implements IControlTemplateStorage.GetTemplateNamesForCategory
            Return templates.Where(Function(x) x.Category = categoryName).Select(Function(y) y.Name).ToArray()
        End Function

        Public Function GetData(ByVal categoryName As String, ByVal templateName As String) As Byte() Implements IControlTemplateStorage.GetData
            Dim item As TemplateLayoutItem = templates.Where(Function(x) x.Category = categoryName AndAlso x.Name = templateName).FirstOrDefault()
            Return If(item Is Nothing, Nothing, item.LayoutBytes)
        End Function

        Public Sub SetData(ByVal categoryName As String, ByVal templateName As String, ByVal templateLayout() As Byte) Implements IControlTemplateStorage.SetData
            Dim item As TemplateLayoutItem = templates.Where(Function(x) x.Category = categoryName AndAlso x.Name = templateName).FirstOrDefault()
            If item Is Nothing Then
                templates.Add(New TemplateLayoutItem() With {.Name = templateName, .Category = categoryName, .LayoutBytes = templateLayout})
            End If
        End Sub

        Private Class TemplateLayoutItem
            Public Property Category() As String
            Public Property Name() As String
            Public Property LayoutBytes() As Byte()
        End Class
    End Class
End Namespace
